### ISSUE TITLE
checkout: consistency in handling files that are missing version info

### ISSUE THREAD

**Issue by efiop** [iterative/dvc#5913]
`dvc checkout`(and `dvc pull` since it uses it internally) will error-out if version-info is missing, while `dvc push` will just warn us, which creates a confusing inconsistency. For example, it makes CML guys use `|| true` for `dvc pull` https://github.com/DavidGOrtega/cml-dvc-test/runs/2438634469 CC @DavidGOrtega , as they don't have `dvc.lock` on initial run.

Current `dvc push` behaviour with warnings is also quite annoying and doesn't scale well, as it might get way to noisy for a big pipeline. So we should probably agree on one common behaviour for all such cases that will make sense to our users. CC @dberenbaum

**Comment by dberenbaum** [comment]
> Current `dvc push` behaviour with warnings is also quite annoying and doesn't scale well, as it might get way to noisy for a big pipeline.

Isn't this what `--quiet` is for? It would be great to have some nice summary option that would be a compromise between no output and overly verbose output, but I think that's separate from the issue here, which is about an inconsistency between `push` and `pull`. If I have done `dvc stage add ... && dvc push`, then `dvc pull` right after should not fail.

**Comment by dberenbaum** [comment]
@efiop Are you waiting on product decisions here? I think we can revisit the warnings, but seems like we can implement pull behavior to match push for now.

**Comment by karajan1001** [comment]
I'm sorry, may I asked which version info is missing

**Comment by dberenbaum** [comment]
@karajan1001 This is the scenario as far as I remember:

* `dvc add` data and commit `.dvc` file.
* `dvc stage add` and commit `dvc.yaml` (the stage does not get run).
* `git push` and `dvc push`, which warns the user that version info is missing for the outputs of the `dvc.yaml` stage.
* On remote server in CML, `git pull` and `dvc pull`, which throws an error because version info is missing for the outputs of the `dvc.yaml` stage.

**Comment by efiop** [comment]
@dberenbaum We might consider starting to treat no-version-info checkout errors as nonfatal. We have some visualization of result in checkout right now (A, D etc), so maybe there is something we can do there instead of push-like warnings, need to take a look.

@skshetry WDYT? ^

**Comment by skshetry** [comment]
I think that the incomplete checkout is a failure, be it due to missing cache or missing version info. `|| true` is the correct way to handle that. We could provide a `--ignore-errors` flags to not error out though.

**Comment by pared** [comment]
Related #6039

**Comment by dberenbaum** [comment]
@skshetry Should push also fail?

In the scenario above, it seems surprising to me that push or pull would fail. Should any stage that hasn't been run yet cause a failure on push or pull?

**Comment by dberenbaum** [comment]
> Should any stage that hasn't been run yet cause a failure on push or pull?

Ping to keep this discussion moving. Any thoughts?

**Comment by skshetry** [comment]
> Ping to keep this discussion moving. Any thoughts?

For the recently created stage, maybe we should not fail, but there could be other reasons why the version info might be missing:
1. git-ignored/dvc-ignored `dvc.lock`,
2. missing/deleted `dvc.lock` file,
3. unreproduced stage,
4. mismatched entries in the `dvc.yaml` and `dvc.lock` file, and
5. missing `md5` hash on `.dvc` file, etc.

**Comment by skshetry** [comment]
> Sure, there could be lots of reasons for not having version info. If there's no way to determine whether this is expected, isn't that what warnings are for?


@dberenbaum, I am not clear on how it should be handled. I understand `dvc checkout` to mean "checkout everything" from the index (i.e. stages). Users might depend on it to exist later in their workspace, so just throwing warnings might make it fail later on the user's script.

 > wouldn't the user's workspace remain untouched anyway? I don't see the danger.

Right now, we delete the file if we are missing version info, even though we fail.

**Comment by dberenbaum** [comment]
> Right now, we delete the file if we are missing version info, even though we fail.

Okay, that does seem more dangerous! Although if it's still happening on failure, then it doesn't seem like this behavior has much to do with whether the command fails or not. Also, why does dvc delete the file?

**Comment by dberenbaum** [comment]
Ping @skshetry

**Comment by skshetry** [comment]
I don't have a good idea on why it deletes the file on missing version info. DVC behaves this in a lot of places, which I'd like to fix (eg: `dvc add` removing stage files on failure, etc). Pinging @efiop.

**Comment by efiop** [comment]
> I don't have a good idea on why it deletes the file on missing version info. 

Just an old assumption so that you are not confused by `checkout` that will keep some random file even though you don't have version info in a corresponding dvcfile. Could reconsider it, sure.

**Comment by pmrowla** [comment]
related #4746

**Comment by dberenbaum** [comment]
@skshetry What do you think about introducing this `--ignore-missing` option or similar  from #4746 to `checkout` and `pull`?

**Comment by agrizzli** [comment]
It would be great to have such an option to `--ignore-missing` outputs (including metrics and plots) during `dvc checkout`! 

In my use case, I define pipeline stages in `dvc.yaml` to be run on a remote cluster. I cannot prepopulate them locally using `dvc repro` (or similar) due to hardware requirements imposed by those stages. The only workaround is to manually create them as empty files and then use `dvc run -f --no-exec` as described [here](https://dvc.org/doc/user-guide/how-to/add-deps-or-outs-to-a-stage), which is quite annoying.

**Comment by dberenbaum** [comment]
Closing this in favor of #4746, but feel free to reopen if there's any other aspects you want to keep discussing.
